packages/core/src/cloud-wallet.ts

```diff
@@ -2,114 +2,337 @@ import {
 
-  const masterKeyLength = mappingData[0];
-  const originalSecondaryKeyLength = mappingData[1];
-  const mappingBytes = mappingData.slice(2);
+  // If there are multiple documents, try each one until one works
+  for (let i = 0; i < result.documents.length; i += 1) {
+    const keyMappingDoc = result.documents[i];
+    const encryptedKey = keyMappingDoc.content.encryptedKey;
 
-  if (mappingBytes.length === 0 || mappingBytes.length !== masterKeyLength) {
-    throw new Error('Invalid key mapping format');
-  }
+    try {
+      const masterKey = await decryptMasterKey(encryptedKey, decryptionKey, iv);
 
-  if (secondaryKeyBytes.length < originalSecondaryKeyLength) {
-    throw new Error('Secondary key is shorter than the one used for mapping creation');
+      return masterKey;
+    } catch (error) {
+      if (i < result.documents.length - 1) {
+        // Try the next document
+        continue;
+      }
+
+      throw new Error('Biometric authentication failed: Invalid biometric data');
+    }
   }
+}
 
-  const masterKeyBytes = new Uint8Array(masterKeyLength);
+/**
+ * Initializes the Cloud Wallet using biometric authentication
+ * @param edvUrl Cloud wallet vault URL
+ * @param authKey Cloud wallet auth key
+ * @param biometricData User's biometric data
+ * @param identifier User's identifier (email, phone number, etc.)
+ * @param dataStore Optional data store for the wallet
+ * @returns Initialized cloud wallet
+ */
+export async function initializeCloudWalletWithBiometrics(
+  edvUrl: string,
+  authKey: string,
+  biometricData: any,
+  identifier: string,
+  dataStore?: any
+): Promise<any> {
+  // Step 1: Authenticate with biometrics to get master key
+  const masterKey = await authenticateWithBiometrics(
+    edvUrl,
+    authKey,
+    biometricData,
+    identifier
+  );
+
+  // Step 2: Initialize the cloud wallet with the master key
+  return initializeCloudWallet({
+    dataStore,
+    edvUrl,
+    authKey,
+    masterKey
+  });
+}
 
-  const overlapLength = Math.min(masterKeyLength, originalSecondaryKeyLength);
-  for (let i = 0; i < overlapLength; i++) {
-    masterKeyBytes[i] = secondaryKeyBytes[i] ^ mappingBytes[i];
-  }
+interface QueuedOperation {
+  operation: () => Promise<any>;
+  resolve: (value: any) => void;
+  reject: (error: any) => void;
+}
 
-  for (let i = overlapLength; i < masterKeyLength; i++) {
-    masterKeyBytes[i] = mappingBytes[i] ^ 0xFF;
-  }
+interface DocumentQueue {
+  operations: QueuedOperation[];
+  isProcessing: boolean;
+}
+
+export async function generateCloudWalletMasterKey(): Promise<{ mnemonic: string; masterKey: string }> {
+  const mnemonic = await utilCryptoService.mnemonicGenerate(MNEMONIC_WORD_COUNT);
 
-  return base64url.encode(Buffer.from(masterKeyBytes));
+  const seedBytes = await utilCryptoService.mnemonicToMiniSecret(mnemonic);
+  const masterKey = base64url.encode(Buffer.from(seedBytes));
```

Comment:
its a waste of time/cpu to constantly encode/decode keys, just pass as buffer/u8 instead of strings

Reply:
This function is user-facing and I don't think returning the master key as buffer is a good DX, but I agree there were some unnecessary encoding/decoding which I fixed